trieste/acquisition/optimizer.py

```diff
@@ -35,24 +34,91 @@
 AcquisitionOptimizer = Callable[[SP, AcquisitionFunction], TensorType]
 """
 Type alias for a function that returns the single point that maximizes an acquisition function over
-a search space. For a search space with points of shape S, and acquisition function with input shape
-[...] + S output shape [..., 1], the :data:`AcquisitionOptimizer` return shape should be [1] + S.
+a search space. For a search space with points of shape [D], and acquisition function with input
+shape [..., B, D] output shape [..., 1], the :const:`AcquisitionOptimizer` return shape should be
+[B, D].
 """
 
 
-@singledispatch
-def optimize(space: Box | DiscreteSearchSpace, target_func: AcquisitionFunction) -> TensorType:
+def automatic_optimizer_selector(
+    space: SearchSpace, target_func: AcquisitionFunction, batch_size: int
+) -> AcquisitionOptimizer[SP]:
     """
+    A wrapper around our :const:`AcquisitionOptimizer`s. This class chooses
+    an :const:`AcquisitionOptimizer` appropriate for the
+    problem's :class:`~trieste.space.SearchSpace` and desired batch size.
+    
     :param space: The space of points over which to search, for points with shape [D].
-    :param target_func: The function to maximise, with input shape [..., D] and output shape
+    :param target_func: The function to maximise, with input shape [..., 1, D] and output shape
         [..., 1].
-    :return: The point in ``space`` that maximises ``target_func``, with shape [1, D].
+    :param batch_size: The number of points in the batch.
+    :return: The batch of points in ``space`` that maximises ``target_func``, with shape [B, D].
     """
 
+    tf.debugging.assert_positive(batch_size)
+
+    if isinstance(space, Box):
+        batch_size_one_optimizer = optimize_continuous
+    elif isinstance(space, DiscreteSearchSpace):
+        batch_size_one_optimizer = optimize_discrete
+    else:
+        raise NotImplementedError(
+            """ No optimizer currentely supports acquisition function
+            maximisation over this search space"""
+        )
+
+    if batch_size == 1:
+        return batch_size_one_optimizer(space, target_func)
+    else:
+        return optimize_batch(batch_size_one_optimizer, space, target_func, batch_size)
+
+
+def optimize_batch(
+    batch_size_one_optimizer: AcquisitionOptimizer,
+    space: SearchSpace,
+    target_func: AcquisitionFunction,
```

Comment:
these arguments are no longer needed. Indeed it's all more composable if we leave them out. As it currently is, users need to specify the search space and acquisition function at the same time as the batch size. This isn't necessary and could create coupling in their code.

Reply:
Yep this makes sense!